MoreLinq/EndsWith.cs

```diff
@@ -71,11 +71,31 @@ public static bool EndsWith<T>(this IEnumerable<T> first, IEnumerable<T> second,
 
             comparer = comparer ?? EqualityComparer<T>.Default;
 
-            var secondCollection = second as ICollection<T> ?? second.ToList();
-            using (var firstIter = first.TakeLast(secondCollection.Count).GetEnumerator())
+            var secondCollection = second;
```

Comment:
I find that this entire section up to the `using` can be improved and folded into a simple expression, using either a tuple:

```c#
            List<T> secondList;
            var (secondCollection, secondCount)
                = second is ICollection<T> collection ? (second, collection.Count)
#if IREADONLY
                : second is IReadOnlyCollection<T> readOnlyCollection ? (second, readOnlyCollection.Count)
#endif
                : (secondList = second.ToList(), secondList.Count);
```

or delegating to a local function:

```c#
            List<T> secondList;
            return second is ICollection<T> collection ? Impl(second, collection.Count)
#if IREADONLY
                 : second is IReadOnlyCollection<T> readOnlyCollection ? Impl(second, readOnlyCollection.Count)
#endif
                 : Impl(secondList = second.ToList(), secondList.Count);

            bool Impl(IEnumerable<T> snd, int count)
            {
                using (var firstIter = first.TakeLast(count).GetEnumerator())
                    return snd.All(item => firstIter.MoveNext() && comparer.Equals(firstIter.Current, item));
            }
```

The flow with both is much easier to follow and reason about.

I would have liked to say take your pick but the first one is a no-go for now due to a regression in Roslyn that unnecessarily forces the use of C# 7.1 (see dotnet/roslyn#21518).